teps/0032-tekton-notifications.md

```diff
@@ -0,0 +1,144 @@
+features that may help implementing notifications:
+
+- [Cloud events](https://tekton.dev/docs/pipelines/events/#events-via-cloudevents)
+- [Finally](https://tekton.dev/docs/pipelines/pipelines/#adding-finally-to-the-pipeline)
+
+Cloud events are triggered every time a pipeline or task is started, begins
+executing and is terminated. Cloud events, combined with Tekton Triggers can
+be used to execute tasks and pipelines as a reaction to an event. This is
+used in our dogfooding cluster to generate notifications for failed CD pipelines.
+It works fine, however it requires a good amount of YAML and plumbing resource
+together to achieve the goal of notifications.
+
+The "Finally" section in a pipeline allows execution tasks once all regular tasks
+completed execution. Such "final" tasks can be used to send notifications about
+the outcome of a pipeline, as they will always be invoked, even if a task in the
+main pipeline failed. However final tasks are only executed at the end, so they
+cannot be used to "start" notifications, and they are limited to pipelines.
+Work is in progress to allow "final" tasks access to the execution status of tasks
+in the main pipeline.
+
+Tekton users would highly benefit from native notification support.
+
+### Goals
+
+The main goal of the TEP is to design and implement a notification system that
+is both easy to use and flexible, for delivering notification targeted to humans.
+
+A secondary goal is to make sure that notification can benefit from our growing
+catalog of tasks, several of which have been written to deliver notification through
+different platforms.
+
+### Non-Goals
+
+- Build a machine to machine communication protocols
+- Build pre/post execution hook for steps, tasks or pipelines
+- Build a catalog of integration with different notification, chat or email systems
+- Build a system to send large amounts of data produced by a task or pipeline
+- Write a controller for cloud events (even if that could be part of the solution)
+
+### Use Cases (optional)
+
+- Notify individuals or groups about the failure of a pipeline
+- Notify individuals or groups about the completion of a long running task
+- Notify individuals or groups about the start or a pipeline that was triggered as
+  a consequence of an automated event (cron, pull request, manual approval)
+
+## Requirements
+
+This section is a stub and subject to change.
+
+Once notifications are implemented, we do not want to force Tekton users to modify
+all their existing tasks or pipelines to enable notifications for them:
+
+- Allow using catalog tasks as is
+- Allow notifications for pipelines with no change to the pipeline itself
+
+For installations that run large number of task and pipelines, the configuration
+of notification should be designed to be scalable and easy to maintain.
+
+- Allow selecting pipelines and tasks to be notified about via selectors
+- Allow selection events to be notified about
+- Allow re-use of notification "rules"
```

Comment:
Mostly task and pipeline authors, but it could also be cluster admins.

One of the things that I find frustrating in `triggers` today is that I have to re-write relatively complex CEL filters in many places and I cannot really re-use them. If one single parameter changes, and/or the bindings and trigger template attached are different, I need a new `trigger` resource.

Thus the idea to generate trigger resources from a resource on top, that can be combined with notification templates that represent bindings + trigger template.

One of the options (maybe I should add it to the alternatives) would be to have new resources defined on triggers side, that allow breaking down the `trigger` object in two and re-use filters. 

I feel that making that more general purpose might be more complex, and I don't know if it there would be a strong enough use case on triggers side to do that. Also I fear that that might drive us away from the focus on notifications specifically.  It's worth discussing anyways.